query/src/main/scala/filodb/query/exec/SelectRawPartitionsExec.scala

```diff
@@ -38,18 +39,39 @@ final case class SelectRawPartitionsExec(id: String,
                                          colIds: Seq[Types.ColumnId]) extends LeafExecPlan {
   import SelectRawPartitionsExec._
 
-  require(colIds.nonEmpty)
-
   protected[filodb] def schemaOfDoExecute(dataset: Dataset): ResultSchema = {
-    val numRowKeyCols = colIds.zip(dataset.rowKeyIDs).takeWhile { case (a, b) => a == b }.length
+    require(dataset.rowKeyIDs.forall(rk => !colIds.contains(rk)),
+      "User selected columns should not include timestamp (row-key); it will be auto-prepended")
+
+    val selectedColIds = selectColIds(dataset)
+    val numRowKeyCols = selectedColIds.zip(dataset.rowKeyIDs).takeWhile { case (a, b) => a == b }.length
 
     // Add the max column to the schema together with Histograms for max computation -- just in case it's needed
     // But make sure the max column isn't already included
-    histMaxColumn(dataset, colIds).filter { mId => !(colIds contains mId) }
+    histMaxColumn(dataset, selectedColIds).filter { mId => !(colIds contains mId) }
                                   .map { maxColId =>
-      ResultSchema(dataset.infosFromIDs(colIds :+ maxColId), numRowKeyCols, colIDs = (colIds :+ maxColId))
+      ResultSchema(dataset.infosFromIDs(selectedColIds :+ maxColId),
+        numRowKeyCols, colIDs = (selectedColIds :+ maxColId))
     }.getOrElse {
-      ResultSchema(dataset.infosFromIDs(colIds), numRowKeyCols, colIDs = colIds)
+      ResultSchema(dataset.infosFromIDs(selectedColIds), numRowKeyCols, colIDs = selectedColIds)
+    }
+  }
+
+  private def selectColIds(dataset: Dataset) = {
+    dataset.rowKeyIDs ++ {
+      if (colIds.nonEmpty) {
+        // query is selecting specific columns
+        colIds
+      } else if (!dataset.hasDownsampledData) {
+        // needs to select raw data
+        colIds ++ dataset.colIDs(dataset.options.valueColumn).get
+      } else {
+        // need to select column based on range function
+        val colNames = rangeVectorTransformers.find(_.isInstanceOf[PeriodicSamplesMapper]).map { p =>
+          RangeFunction.downsampleColsFromRangeFunction(dataset, p.asInstanceOf[PeriodicSamplesMapper].functionId)
```

Comment:
I'm thinking philosophically that instead of embedding the choice of columns here in the execution phase, it is better to select it automatically in the query planning phase.  However in the future one might not know it is downsampled data until the execution phase.  Just that it keeps everything cleaner at the lower level, because now we have multiple places where this logic is.  Also should not need to automatically add rowKeyIDs here, it might already have been added before.

Reply:
Hence removed it everywhere else but here. 

Fully agree that it should happen in one place. I took the "here" approach as against query planning.
Hence column selection was removed from everywhere else and placed right here. 

That is the reason there is the assertion "require rowKeys to not be present".

It is debatable where column selection needs to happen: here during runtime, or at planning time. Let's defer that decision to the phase when we address queries for multi-schema. A lot more information needs to go into that decision making. For now, I have colocated all relevant code in one place. If it needs to be moved elsewhere, it can happen in a later PR.

Reply:
👍